### PR TITLE
pallet-ibc: add whitelisting

### PR DESCRIPTION
**Pull request by blasrodri**
This feature is useful while testing the DOT<>KSM bridge on our parachains. Will likely be removed.

EDIT:
A simple implementation would be to have some harcoded addresses both on Picasso and Composable
to simply test the bridge without opening it up to the masses.

### CODE REVIEW

**Review by Wizdave97 — COMMENTED** [review]
The whitelist should be used in the `transfer` extrinsic.

**Review by dzmitry-lahoda — CHANGES_REQUESTED** [review]
please use Origin `ensure`

**Comment by blasrodri** [comment]
> please use Origin `ensure`

where?

**Comment by blasrodri** [comment]
@dzmitry-lahoda do you mean

```rs
		type Whitelist: Contains<<Self as frame_system::Config>::AccountId>
			+ EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
```

**Comment by dzmitry-lahoda** [comment]
yes. No Contains. Just Ensure. Set it to SignedOrRoot. Or just EnsureSigned

**Comment by dzmitry-lahoda** [comment]
you do not need Contains.

**Comment by blasrodri** [comment]
> you do not need Contains.

but if we want to create a white list mechanism, how does this work?

**Comment by dzmitry-lahoda** [comment]
ensure is abstraction over contains. it can do contains. ensure is just contains for origins.

**Comment by dzmitry-lahoda** [comment]
so just make ensure, and in dali will setup proper config.

**Comment by Wizdave97** [comment]
I think `Contains` is more appropriate in this case. The goal is to ensure the origin executing the transaction is part of a whitelisted set

**Comment by Wizdave97** [comment]
> I think `Contains` is more appropriate in this case. The goal is to ensure the origin executing the transaction is part of a whitelisted set

@dzmitry-lahoda  the goal is just to have a few accounts hardcoded in the runtime config that are allowed to use ibc transfers.

**Comment by blasrodri** [comment]
@KaiserKarel what's your opinion here?

File: utils/parachain-node/runtime/src/lib.rs [hunk 1]
```diff
@@ -698,8 +698,16 @@ impl pallet_ibc::Config for Runtime {
 	type AdminOrigin = EnsureRoot<AccountId>;
 	type SentryOrigin = EnsureRoot<AccountId>;
 	type SpamProtectionDeposit = SpamProtectionDeposit;
+	type Whitelist = SimpleIBCWhiteList;
```
Comment (KaiserKarel) [inline utils/parachain-node/runtime/src/lib.rs:701]: Can be the AllowAll type for dedup

**Review by KaiserKarel — DISMISSED** [review]

**Review by Wizdave97 — APPROVED** [review]

**Comment by blasrodri** [comment]
@dzmitry-lahoda can you please review it again?

File: contracts/pallet-ibc/src/lib.rs [hunk 2]
```diff
@@ -582,6 +586,10 @@ pub mod pallet {
 			amount: T::Balance,
 		) -> DispatchResult {
 			let origin = ensure_signed(origin)?;
+			// Ensure that the signer is whitelisted
+			if !T::Whitelist::contains(&origin) {
+				return Err(Error::<T>::AccessDenied.into())
+			}
```
[the remark this answers is not in the document]
Comment (blasrodri) [reply contracts/pallet-ibc/src/lib.rs:606]: I had to revert, because of compilation issues

**Review by obsessed-cake — APPROVED** [review]

**Pull request merged**